ctk/common/src/main/kotlin/org/codice/compliance/verification/profile/SingleSignOnProfileVerifier.kt

```diff
@@ -111,102 +110,93 @@ class SingleSignOnProfileVerifier(private val response: Node,
+    }
+
+    @Suppress("ComplexCondition")
+    private fun verifyBearerSubjectConfirmations(bearerSubjectConfirmations: List<Node>) {
         if (bearerSubjectConfirmations
                         .filter { it.children("SubjectConfirmationData").isNotEmpty() }
                         .flatMap { it.children("SubjectConfirmationData") }
                         .none {
                             it.attributes.getNamedItem("Recipient").textContent == acsUrl &&
-                                    it.attributes.getNamedItem("InResponseTo").textContent == ID &&
                                     it.attributes.getNamedItem("NotOnOrAfter") != null &&
-                                    it.attributes.getNamedItem("NotBefore") == null
-                        }) {
+                                    it.attributes.getNamedItem("NotBefore") == null &&
+                                    it.attributes.getNamedItem("InResponseTo").textContent ==
+                                    ID }) {
             throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_h,
                     message = "There were no bearer SubjectConfirmation elements that matched " +
                             "the criteria below.",
                     node = response)
         }
     }
 
-    private fun verifyHasSubject(assertion: Node) {
-        if (assertion.children(SUBJECT).size != 1) {
-            throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_f,
-                    message = "${assertion.children(SUBJECT).size} Subject elements were found.",
+    private fun verifyAssertionsHaveAuthnStmts(bearerAssertions: List<Node>) {
+        if (bearerAssertions.none {
+                    it.children(AUTHN_STATEMENT).isNotEmpty()
+                })
+            throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_i,
+                    message = "A Bearer Assertion with an AuthnStatement was not found.",
                     node = response)
+    }
+
+    private fun verifySingleLogoutIncludesSessionIndex(bearerAssertions: List<Node>) {
+        if (idpMetadata.descriptor == null) return
+
+        if (idpMetadata.descriptor?.singleLogoutServices?.isNotEmpty() == true) return
+
+        bearerAssertions.filter { it.children(AUTHN_STATEMENT).isEmpty() }
+                .flatMap { it.children(AUTHN_STATEMENT) }
+                .forEach {
+                    if (it.attributes.getNamedItem("SessionIndex") == null) {
+                        throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_j,
+                                message = "Single Logout support found in IdP metadata, but no " +
+                                        "SessionIndex was found.",
+                                node = response)
+                    }
+                }
+    }
+
+    private fun verifyBearerAssertionsContainAudienceRestriction(bearerAssertions: List<Node>) {
+        bearerAssertions.forEach {
+            if (it.children("Conditions").isNotEmpty()) {
```

Comment:
✏️ Map to `it.children("Conditions")` and then filter on nonEmpty?
And maybe add a method to `extractAudience` so you can do another map call, terminating with an any or none to check audience against `SP_ISSUER`?